### PR TITLE
Table profile added

### LINKED ISSUE #66: proposal: `%sqlcmd profile`
When working with a new dataset, practitioners need to explore and summarize it quickly: column values, types, distributions, etc. We could create a `%sqlcmd profile` magic that produces an HTML table/report of a table. Similar to [pandas-profiling](https://github.com/ydataai/pandas-profiling), except this would run the analysis on the SQL engine, making it more scalable.

Examples:

display an embedded summary:

```
%sqlcmd profile --table my_table
```

store a report:

```
%sqlcmd profile --table my_table --output report.html
```

### PR DESCRIPTION
**Pull request by yafimvo**
## Describe your changes
`profile` magic command added to `%sqlcmd`.

How to use it:
`%sqlcmd profile -t table_name`

## Issue ticket number and link
Closes #66 

## Checklist before requesting a review

- [X] I have performed a self-review of my code
- [X] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [X] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--168.org.readthedocs.build/en/168/

### CODE REVIEW

**Comment by edublancas** [comment]
is this ready for review?

**Comment by yafimvo** [comment]
Yes

File: doc/_toc.yml [hunk 1]
```diff
@@ -13,6 +13,7 @@ parts:
     - file: compose
     - file: user-guide/tables-columns
     - file: plot-legacy
+    - file: user-guide/explore-tables
```
Comment (edublancas) [inline doc/_toc.yml:16]: let's rename this to data profiling since that's a common term

File: doc/user-guide/explore-tables.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+* The frequency of your top value
+* The mean, standard deviation, min and max values
+* The percentiles of your data: 25%, 50% and 75%.
+
+
+## Examples
+
+### Simple example with SQLite
+
+```{code-cell} ipython3
```
Comment (edublancas) [inline doc/user-guide/explore-tables.md:32]: let's switch this to duckdb with the taxi data to simplify setup: https://jupysql.ploomber.io/en/latest/plot.html

but mention that this is compatible with all major databases

File: doc/user-guide/explore-tables.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+%sqlcmd profile -t track
+```
+
+### Saving report as HTML
+
+To save the generated report as an HTML file, use the `--output`/`-o` attribute followed by the desired file name
+
+```
+%sqlcmd profile -t track --output my-report.html
```
Comment (edublancas) [inline doc/user-guide/explore-tables.md:95]: can we embed the sample report here? we can use

```python
from IPython.display import HTML
HTML("my-report.html")
```

File: src/sql/inspect.py [hunk 4]
```diff
@@ -73,6 +74,147 @@ def __init__(self, name, schema, conn=None) -> None:
...
+            table_stats[column]["freq"] = result_col_freq_values.dict()["frequency"][0]
+            table_stats[column]["unique"] = result_col_unique_values.dict()[
+                "unique_count"
+            ][0]
+            table_stats[column]["top"] = result_col_freq_values.dict()[column][0]
+            table_stats[column]["min"] = col_values[0]
+            table_stats[column]["max"] = col_values[count - 1]
+
+            try:
+                mean = sum(col_values) / count
```
Comment (edublancas) [inline src/sql/inspect.py:152]: computing the stats this way won't scale, we need tot compute them with SQL directly (we can compute percentiles with `percentile_disc`), check out the plotting module, the code will be similar

File: src/sql/magic_cmd.py [hunk 5]
```diff
@@ -33,6 +34,18 @@ def error(self, message):
...
+    displaycon = True
+    autolimit = None
+    style = "DEFAULT"
+    short_errors = True
+    displaylimit = None
+    autopandas = False
+    column_local_vars = False
+    feedback = False
+    autocommit = False
+    autopolars = False
```
Comment (edublancas) [inline src/sql/magic_cmd.py:46]: are these needed? these look like the config from the original sqlmagic and I don't think we're using them here

File: src/sql/magic_cmd.py [hunk 6]
```diff
@@ -33,6 +34,18 @@ def error(self, message):
...
+
+    @needs_local_scope
```
Comment (edublancas) [inline src/sql/magic_cmd.py:48]: same, do we need this?

File: src/sql/magic_cmd.py [hunk 7]
```diff
@@ -65,8 +78,34 @@ def execute(self, line="", cell="", local_ns=None):
 
             args = parser.parse_args(others)
             return inspect.get_columns(name=args.table, schema=args.schema)
+
+        elif cmd_name == "profile":
+            parser = CmdParser()
+
+            parser.add_argument(
```
Comment (edublancas) [inline src/sql/magic_cmd.py:85]: let's add support for --schema. the command that lists tables and columns supports it, you can use it as a guide

**Review by edublancas — CHANGES_REQUESTED** [review]

**Comment by edublancas** [comment]
I also noticed that the numbers in the profiling table are too long. let's implement a custom format display that shortens it by displaying them in scientific notation. @neelasha23 implemented something like that for sklearn-evaluation's interactive confusion matrix, maybe we can re-use the code?

File: src/sql/inspect.py (hunk 4 above)
Comment (yafimvo) [reply src/sql/inspect.py:152]: `percentile_disc` won't work on sqlite. How do we want to handle it?

Currently, if we can't calculate the Nth percentile due to non-numeric value - we show 'nan' in the Nth percentile cell.

I can hide the Nth percentile cell if the database doesn't support the `percentile_disc.

File: src/sql/magic_cmd.py (hunk 5 above)
Comment (yafimvo) [reply src/sql/magic_cmd.py:46]: I tried to pass `self.config` but it looks like there is no default configuration there and it fails to run the query.
e.g: `self.config.autocommit` returns `LazyConfigValue` (which implies that there is no such config attribute?)

I noticed that if we run `%sqlcmd columns` on a table loaded with DuckDB we get `NoSuchTableError`.

e.g:
```
%sqlcmd columns --table "yellow_tripdata_2021-01.parquet"
```

Is this may be related?

*Note
This is why I'm using SQL query to get column names instead `inspector.get_columns(name, schema)`.

File: src/sql/magic_cmd.py (hunk 6 above)
Comment (yafimvo) [reply src/sql/magic_cmd.py:48]: I need the `local_ns` in order to use `run(conn, sql, config, user_namespace)`. If I remove this decorator I don't have it.

**Comment by yafimvo** [comment]
> I also noticed that the numbers in the profiling table are too long. let's implement a custom format display that shortens it by displaying them in scientific notation. @neelasha23 implemented something like that for sklearn-evaluation's interactive confusion matrix, maybe we can re-use the code?

I found this code. It consists of 2 parts, one (`convert_to_scientific`) takes data in a specific format (key-value) and checks if it's a number. The second part (`_is_long_number`) checks its length and formats it using `np`. 

We can use `_is_long_number` as is but `convert_to_scientific` is a bit different. In the meantime, I added both of them to `util` in `jupysql` (with some modifications to `convert_to_scientific`).

@edublancas 
Do you think we should move `_is_long_number` to ploomber_core?

File: setup.py [hunk 8]
```diff
@@ -24,6 +24,7 @@
     "jinja2",
     "ploomber-core>=0.2.4",
     'importlib-metadata;python_version<"3.8"',
+    "numpy"
```
Comment (edublancas) [inline setup.py:27]: looks like we can remove this since we're running things with SQL

File: src/sql/inspect.py [hunk 9]
```diff
@@ -73,6 +75,170 @@ def __init__(self, name, schema, conn=None) -> None:
...
+    Freq - Frequency of the top value
+
+    """
+
+    def __init__(self, table_name, schema=None, config=None, user_ns=None) -> None:
+
+        if schema:
+            table_name = f"{schema}.{table_name}"
+
+        result_table_columns = sql.run.run(
```
Comment (edublancas) [inline src/sql/inspect.py:110]: using run might break since it's output is affected by the user's settings, let's create a `run_raw` function that calls `conn.session.execute` and returns the results without any extra logic (see the implementation of `run`)

File: src/sql/inspect.py (hunk 4 above)
Comment (edublancas) [reply src/sql/inspect.py:157]: yeah, I think it's better to hide it - let's include a message at the end of the table saying that the function isn't supported so the user knows why it doesn't show up

File: src/sql/inspect.py [hunk 10]
```diff
@@ -73,6 +75,170 @@ def __init__(self, name, schema, conn=None) -> None:
...
+                Connection.current,
+                f"""SELECT {column} FROM {table_name} WHERE {column}
+                IS NOT NULL AND TRIM({column}) <> ''
+                ORDER BY {column} ASC
+                """,
+                config,
+                user_ns,
+            )
+
+            col_values = result_non_empty_values.dict()[column]
```
Comment (edublancas) [inline src/sql/inspect.py:145]: looks like this is fetching all the data right? this is unfeasible as users might profile tables with millions of rows. we should use `AVG` for the mean, and `COUNT(*)` for counting; the idea is that we should only fetch the aggregated data to the client, everything else should happen in the database

File: src/sql/inspect.py [hunk 11]
```diff
@@ -73,6 +75,170 @@ def __init__(self, name, schema, conn=None) -> None:
...
+            count = len(col_values)
+            table_stats[column]["count"] = count
+            table_stats[column]["freq"] = result_col_freq_values.dict()["frequency"][0]
+            table_stats[column]["unique"] = result_col_unique_values.dict()[
+                "unique_count"
+            ][0]
+            table_stats[column]["top"] = result_col_freq_values.dict()[column][0]
+            table_stats[column]["min"] = col_values[0]
+            table_stats[column]["max"] = col_values[count - 1]
```
Comment (edublancas) [inline src/sql/inspect.py:154]: same here, we should use `MIN` and `MAX`

File: src/sql/inspect.py [hunk 12]
```diff
@@ -73,6 +75,170 @@ def __init__(self, name, schema, conn=None) -> None:
...
+
+            try:
+                mean = sum(col_values) / count
+                table_stats[column]["mean"] = mean
+
+                values_sum = sum([(math.pow((v - mean), 2)) for v in col_values])
+                std = math.sqrt(values_sum / (count - 1))
```
Comment (edublancas) [inline src/sql/inspect.py:161]: we can use `STDEV` here; if the database doesn't support it, we can show NAN

File: src/sql/magic_cmd.py (hunk 5 above)
Comment (edublancas) [reply src/sql/magic_cmd.py:46]: Ok, I think this will be fixed if we migrate to a `run_raw` function as described in my earlier comment

> I noticed that if we run %sqlcmd columns on a table loaded with DuckDB we get NoSuchTableError.

I think this is expected since the parquet is not a table; but can you open an issue? I think we should document it since it might confuse users

File: src/sql/magic_cmd.py (hunk 6 above)
Comment (edublancas) [reply src/sql/magic_cmd.py:48]: same, if we move to a `run_raw` function, we won't need this anymore

**Review by edublancas — CHANGES_REQUESTED** [review]

**Comment by edublancas** [comment]
> I found this code. It consists of 2 parts, one (`convert_to_scientific`) takes data in a specific format (key-value) and checks if it's a number. The second part (`_is_long_number`) checks its length and formats it using `np`.
> 
> We can use `_is_long_number` as is but `convert_to_scientific` is a bit different. In the meantime, I added both of them to `util` in `jupysql` (with some modifications to `convert_to_scientific`).
> 
> @edublancas Do you think we should move `_is_long_number` to ploomber_core?

the dependence on numpy is a problem here. it sounds like too much to add numpy just to use such function.  looks like with can do it without numpy as well: https://stackoverflow.com/a/69569277/709975

let's keep it here for now, we'll move it to core if we need it elsewhere

File: src/sql/inspect.py [hunk 13]
```diff
@@ -73,6 +75,145 @@ def __init__(self, name, schema, conn=None) -> None:
...
+
+            except TypeError:
+                # for non numeric values
+                table_stats[column]["mean"] = math.nan
+                table_stats[column]["std"] = math.nan
+                table_stats[column]["25%"] = math.nan
+                table_stats[column]["50%"] = math.nan
+                table_stats[column]["75%"] = math.nan
+
+            except BaseException:
```
Comment (edublancas) [inline src/sql/inspect.py:191]:
```suggestion
            except Exception:
```

File: src/sql/run.py [hunk 14]
```diff
@@ -106,13 +106,18 @@ def __init__(self, sqlaproxy, config):
         self.keys = {}
         if sqlaproxy.returns_rows:
             self.keys = sqlaproxy.keys()
-            if config.autolimit:
+            if isinstance(config.autolimit, int) and config.autolimit > 0:
```
Comment (edublancas) [inline src/sql/run.py:109]: why do we need this? this would break existing compatibility (setting None and setting autolimit to 0 should display all values) - it's a bit counterintuitive but we inherited this behavior from ipython-sql

File: src/sql/run.py [hunk 15]
```diff
@@ -106,13 +106,18 @@ def __init__(self, sqlaproxy, config):
...
                 list.__init__(self, sqlaproxy.fetchall())
             self.field_names = unduplicate_field_names(self.keys)
+
+            _style = None
+            if isinstance(config.style, str):
+                _style = prettytable.__dict__[config.style.upper()]
+
             self.pretty = PrettyTable(
-                self.field_names, style=prettytable.__dict__[config.style.upper()]
+                self.field_names, style=_style
```
Comment (edublancas) [inline src/sql/run.py:120]: why was this changed?

File: src/tests/test_magic_cmd.py [hunk 16]
```diff
@@ -70,3 +72,117 @@ def test_columns_with_schema(ip, tmp_empty):
     ).result._repr_html_()
 
     assert "some_number" in out
+
+
+def test_table_profile(ip, tmp_empty):
```
Comment (edublancas) [inline src/tests/test_magic_cmd.py:77]: can you add some tests to the integration testing file? we should check if this works with other databases, we know it doesn't work with sqlite so we can ignore it.

but we should check for the other ones, we can mark tests as xfail fo the ones that don't pass the tests and we'll fix them later

**Review by edublancas — CHANGES_REQUESTED** [review]
how should I interpret an empty table cell vs a cell with nan?

in the tutorial, I see this:

image

File: src/sql/run.py (hunk 14 above)
Comment (yafimvo) [reply src/sql/run.py:109]: Since I removed the hardcoded configuration from `SqlCmdMagic` there are some missing default configurations (`autolimit` and `style`). In this case, `config.autolimit` and `config.style` return `<LazyConfigValue>`.

I added the `config.autolimit > 0` since according to one of the tests if autolimit is 0 we should return everything.

**Comment by yafimvo** [comment]
> how should I interpret an empty table cell vs a cell with nan?
> 
> in the tutorial, I see this:
> 
> image

It happened since it tried to run `stdev` on non-numeric values (DateTime) and the exception was vague (`sqlalchemy.exc.ProgrammingError`), so it ignored these cells.

Changed it.

File: src/tests/test_magic_cmd.py (hunk 16 above)
Comment (yafimvo) [reply src/tests/test_magic_cmd.py:77]: Added test for each database but with the relevant profile fields (DuckDB and PostgreSQL should work with all fields)

File: src/sql/run.py (hunk 14 above)
Comment (edublancas) [reply src/sql/run.py:109]: ah, I see what you're saying! we have a design problem here.

the configuration is attached to the `%sql` magic, but ideally we want the config to be accessible to all magics. I'm unsure if this is possible so I think for now the best thing to do is to create another version of `run` that doesn't take the config argument (we won't be able to provide the autolimit feature when running the profiling but that's fine)

I remember suggesting creating a `raw_run` function already but I can't remember if it was in the ggplot PR or in a different one that @tonykploomber is working on.

File: src/sql/run.py (hunk 15 above)
Comment (edublancas) [reply src/sql/run.py:120]: looks like this is fixing a bug (since the _style variable wasn't used?)

**Review by edublancas — CHANGES_REQUESTED** [review]
added some comments

**Review by edublancas — APPROVED** [review]
